bip-hashrate-escrows.mediawiki

```diff
@@ -0,0 +1,469 @@
+
+A Hashrate Escrow resembles a 2-of-3 multisig escrow, where the 3rd party (who will arbitrate any disputes) is a decentralized group of people: the dynamic-membership set of Bitcoin Miners. However, the 3rd party does not sign escrow-withdrawal transactions with a private key. Instead, these are "signed" by directing hashpower over them for a period of time.
+
+This project has [http://www.drivechain.info/ a website] which includes [http://www.drivechain.info/faq/index.html a FAQ].
+
+
+==Motivation==
+
+In practice these escrows are likely to be "asymmetric sidechains" of Bitcoin (such as [http://www.rsk.co/ Rootstock]) or "virtual chains" within Bitcoin (such as [https://github.com/blockstack/virtualchain proposed by Blockstack] in mid-2016).
+
+Sidechains have many potential benefits, including:
+
+1. Protect Bitcoin from competition from altcoins and spinoffs. Safely allow competing implementations (of *sidechains*).
+2. Protect Bitcoin from hard fork campaigns. (Such campaigns represent an existential threat to Bitcoin, as well as an avenue for developer corruption.)
+3. Help with review, by making it much easier for reviewers to ignore bad ideas.
+4. Provide an avenue for good-but-confusing ideas to prove their value safely.
+
+
+
+==Specification==
+
+==== Components ====
+
+Hashrate Escrows are built of two types of component: [1] new databases, and [2] new message-interpretations.
+
+===== 1. New Databases =====
+
+* D1. "Escrow_DB" -- a database of "accounts" and their attributes.
+* D2. "Withdrawal_DB" -- a database of pending withdrawals from these accounts, and their statuses.
+
+Please note that these structures (D1 and D2) will not literally exist anywhere in the blockchain. Instead they are constructed from messages...these messages, in contrast, *will* exist in the blockchain (with the exception of M4). 
+
+===== 2. New Messages =====
+
+* M1. "Propose New Escrow"
+* M2. "ACK Escrow Proposal"
+* M3. "Propose Withdrawal"
+* M4. (implied) "ACK Withdrawal"
+* M5. "Execute Deposit"   -- a transfer of BTC from-main-to-side
+* M6. "Execute Withdrawal" -- a transfer of BTC from-side-to-main
+
+
+==== On the Resource Requirements of New Databases ====
+
+The "new" databases are simply reinterpretations of data that are already contained elsewhere in the blockchain. Specifically, M1 M2 and M3 are all located in the block's coinbase txn, and M5 and M6 might be found in any regular txn. M4 is a special case and does not actually need to be included anywhere, so it is not. If you like, you can imagine that the M4s reside in an optional extension block.
+
+In other words, we just rearrange what is already there. Because of this, even though "new databases" are created and stored in memory, the existing bandwidth and storage limits are respected (although, see "M4" below).
```

Comment:
This does not seem to logically follow. The existing databases are not being apparently reduced in size to compensate, so it is strictly adding requirements here, no?

I think it may be appropriate to increase the block weight when these new databases are affected.

Reply:
I suppose that you are right that the "storage" requirements would in some sense increase. But, as a soft fork, this would only be for users who upgrade.

This would also be true if, for example, the latest Bitcoin Core were larger in size, and so downloading and building the software required more disk space. Or it would be true for a release that required the upgrading users to track Segwit and nonSegwit outputs and sort among these (or even, CheckLocktimeVerify outputs and non-CLTV outputs). In such cases I would find it strange to decrease the block size/weight limits to offset this cost.

But nonetheless I think I could remove the "and storage" phrase from the last sentence.

Reply:
Non-upgraded nodes is never a viable option for any softfork. Old nodes remain compatible, but nobody is supposed to actually *use* them.

Reply:
Hmmm, ok.

So, you would like it to be such that, when there are many sidechains and many withdrawals at once, (ie, when M4 is large), then the blockweight limit decreases by this size?

So, for example, the extreme attacker case, where 256 sidechains have been added, and all of them are doing nothing but spamming full node miners all day, such that in each block a new withdrawal is attempted on each sidechain, such that there are 26300*256 withdrawals at once (this would already be comically irrational for each miner, as it would consume a ridiculous amount of regular blockchain space, that they could instead use to include txns and collect txn fees). M4 might need to be as large as 841,670 bytes. So you would want the blocksize/weight to decrease by 842 kb while the network is in such a state? Because, to you, my alternative (that users just disregard what is happening amonst miners) is not rigorous enough?

I could shrink that 842 considerably. One way would be to allow ACK counts to go negative, and then further to have some event trigger upon an ACK-count reaching -250 or so. Upon such a count we could delete all withdrawals (from that sidechain) that have a negative ACK count at that time. Then even with 256 sidechains, M4's size would never need to be more than 8 kb or so.

Reply:
Anything that adds to the burden of nodes doing FULL validation, should count as weight in some way, at least.